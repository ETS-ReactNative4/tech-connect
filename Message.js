import React, { Component } from 'react'
import { StyleSheet, View, Image, Text } from 'react-native'

export default class Message extends Component {
  render() {
    return (
      <View style={ styles.messageContainer }>
        <View style={ styles.imageContainer }>
          <Image source={ require('./profile-pic.jpeg') } style={ styles.userPic }/>
        </View>
        <Text style={ styles.date }>10/10/2019</Text>
        <View style={ styles.messageInfo }>
          <Text style={ styles.name }>Howard Stearn</Text>
          <Text style={ styles.message }>This is Howard's message</Text>
        </View>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  messageContainer: {
    display: 'flex',
    flexDirection: 'row',
    padding: 20,
    borderBottomWidth: 0.5,
    borderBottomColor: 'gray',
  },
  imageContainer: {
    shadowOffset: {  width: 2,  height: 2 },
    shadowRadius: 4,
    shadowColor: 'black',
    shadowOpacity: .2,
  },
  userPic: {
    height: 50,
    width: 50,
    borderRadius: 25
  },
  date: {
    position: 'absolute',
    right: 20,
    top: 20
  },
  name: {
    fontSize: 18
  },
  message: {
    marginTop: 10,
    color: 'gray'
  },
  messageInfo: {
    marginLeft: 8
  }
})