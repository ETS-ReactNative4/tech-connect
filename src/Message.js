import React, { Component } from 'react'
import { StyleSheet, View, Image, Text, TouchableOpacity } from 'react-native'
import { connect } from 'react-redux';


export class Message extends Component {
  constructor() {
    super()
    this.state = {
      showMessage: false
    }
  }

  render() {
    const { message, user } = this.props
    const connection = (user.name === message.receiver.name) ? message.sender.name :  message.receiver.name
    const sentDate = new Date(message.created_at).toUTCString().split(' ').slice(0, 4).join(' ')
    const meetingDate = new Date(message.meeting_date).toUTCString().split(' ').slice(0, 4).join(' ')
    const time = new Date(message.meeting_date).toUTCString().split(' ')[4]
    let messageText
    let messagePhoto
    let messageEnd = (message.meeting_location !== 'N/A') ? `at ${message.meeting_location} on ${meetingDate} at ${time}` : null

    if (message.receiver.name === user.name) {
      messageText = <Text style={ styles.messageBody } >You have { message.status } a meeting with { message.sender.name } { messageEnd }</Text>
      messagePhoto = message.sender.photo
    } else {
      messageText = <Text style={ styles.messageBody } >{ message.receiver.name } has { message.status } a meeting with you { messageEnd }</Text>
      messagePhoto = message.receiver.photo
    }
    return (
      <View style={ styles.messageContainer }>
        <Text style={ styles.date }>{ sentDate }</Text>
        <TouchableOpacity style={ styles.messageInfoContainer } onPress={() => this.setState({showMessage: !this.state.showMessage})}>
          <View style={ styles.imageContainer }>
            <Image source={ {uri: messagePhoto} } style={ styles.userPic }/>
          </View>
          <View style={ styles.messageInfo }>
            <Text style={ styles.name }>{ connection }</Text>
            <Text style={ styles.message }>{ message.status }</Text>
            {
              !this.state.showMessage ? null : messageText
            }
          </View>
        </TouchableOpacity>
      </View>
    )
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
})

export default connect(mapStateToProps)(Message)


const styles = StyleSheet.create({
  messageContainer: {
    display: 'flex',
    flexDirection: 'row',
    padding: 20,
    borderBottomWidth: 0.5,
    borderBottomColor: 'gray',
  },
  messageInfoContainer: {
    display: 'flex',
    flexDirection: 'row',
    maxWidth: '100%',
    marginRight: 10
  },
  imageContainer: {
    shadowOffset: {  width: 2,  height: 2 },
    shadowRadius: 4,
    shadowColor: 'black',
    shadowOpacity: .2,
  },
  userPic: {
    height: 50,
    width: 50,
    borderRadius: 25
  },
  date: {
    position: 'absolute',
    right: 20,
    top: 20
  },
  name: {
    fontSize: 18
  },
  message: {
    marginTop: 5,
    color: 'gray',
  },
  messageBody: {
    marginRight: 10
  },
  messageInfo: {
    marginLeft: 10,
    marginRight: 30,
    marginTop: 5
  }
})