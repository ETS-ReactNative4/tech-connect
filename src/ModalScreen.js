import React, { Component } from 'react'
import { StyleSheet, Text, View, TextInput, Button, TouchableHighlight, TouchableOpacity } from 'react-native';
import { Input } from 'react-native-elements'
import DatePicker from 'react-native-datepicker'
import { connect } from 'react-redux'
import { sendConnectionRequest } from '../apiCalls'
import Icon from 'react-native-vector-icons/Feather'


export class ModalScreen extends Component {
  constructor() {
    super()
    this.state = {
      response: '',
      inputs: {
        datetime_1: '', 
        datetime_2: '', 
        datetime_3: '',
        meeting_1: '',
        meeting_2: '',
        meeting_3: ''   
      }
    }
  }

  sendRequest = async () => {
    const user = this.props.navigation.getParam('user')
    const response = await sendConnectionRequest(this.props.user.api_key, user.id, this.state.inputs)
    this.setState({ response })
  }

  render() {
    const { response, inputs } = this.state
    const { datetime_1, datetime_2, datetime_3, meeting_1, meeting_2, meeting_3 } = this.state.inputs
    const user = this.props.navigation.getParam('user')
    const note = `${user.name} will receive an email with your suggestions. Once confirmed you will receive a notification email.`
    const error = response.length ? <Text style={{color: 'red'}}>{response}</Text> : null
    if (response === 'Email Sent') {
      return (
        <View style={styles.container}>
          <View style={{flexDirection: 'row', alignItems: 'center', justifyContent: 'center'}}>
            <Text style={styles.title}>{ response }</Text>
            <Icon
              name='check-circle'
              size={35}
              color='#fff'
              style={{marginLeft: 5}}
            />
          </View>
          <TouchableHighlight style={styles.btnContainer}>
            <Button
              onPress={() => this.props.navigation.goBack()}
              color={'#fff'}
              title="Done"
            />
          </TouchableHighlight>
        </View>
      )
    } else {
      return (
        <View style={styles.container}>
          <Text style={styles.title}>Request a Meeting</Text>
          <Text style={styles.note}>{ note }</Text>
          <Text style={styles.suggest}>Suggest Three Date/Times:</Text>
          <DatePicker
            style={styles.date}
            date={datetime_1}
            mode='datetime'
            format='YYYY-MM-DD HH:MM:SS'
            minDate={new Date()}
            maxDate='2020-01-01'
            confirmBtnText='Confirm'
            cancelBtnText='Cancel'
            customStyles={{
              dateIcon: {
                position: 'absolute',
                left: 0,
                top: 4,
                marginLeft: 0
              },
              dateInput: {
                marginLeft: 36,
                borderWidth: 0
              }
            }}
            onDateChange={date => this.setState({inputs: {...this.state.inputs, datetime_1: date}})}
          />
          <DatePicker
            style={styles.date}
            date={datetime_2}
            mode='datetime'
            format='YYYY-MM-DD HH:MM:SS'
            minDate={new Date()}
            maxDate='2020-01-01'
            confirmBtnText='Confirm'
            cancelBtnText='Cancel'
            customStyles={{
              dateIcon: {
                position: 'absolute',
                left: 0,
                top: 4,
                marginLeft: 0
              },
              dateInput: {
                marginLeft: 36,
                borderWidth: 0
              }
            }}
            onDateChange={date => this.setState({inputs: {...this.state.inputs, datetime_2: date}})}
          />
          <DatePicker
            style={styles.date}
            date={datetime_3}
            mode='datetime'
            format='YYYY-MM-DD HH:MM:SS'
            minDate={new Date()}
            maxDate='2020-01-01'
            confirmBtnText='Confirm'
            cancelBtnText='Cancel'
            customStyles={{
              dateIcon: {
                position: 'absolute',
                left: 0,
                top: 4,
                marginLeft: 0
              },
              dateInput: {
                marginLeft: 36,
                borderWidth: 0
              }
            }}
            onDateChange={date => this.setState({inputs: {...this.state.inputs, datetime_3: date}})}
          />
          <Text style={styles.suggest}>Suggest Three Locations:</Text>
          <Input 
            onChangeText={text => this.setState({inputs: {...this.state.inputs, meeting_1: text}})}
            containerStyle={styles.input} 
            inputContainerStyle={styles.inputContainer} 
          />
          <Input 
            onChangeText={text => this.setState({inputs: {...this.state.inputs, meeting_2: text}})} 
            containerStyle={styles.input} 
            inputContainerStyle={styles.inputContainer} 
          />
          <Input 
            onChangeText={text => this.setState({inputs: {...this.state.inputs, meeting_3: text}})}
            containerStyle={styles.input} 
            inputContainerStyle={styles.inputContainer} 
          />
          { error }
          <TouchableHighlight style={styles.btnContainer}>
            <Button 
              onPress={this.sendRequest}
              title="Send Request"
              color={'#fff'}
            />
          </TouchableHighlight>
          <TouchableHighlight style={styles.btnContainer}>
            <Button
              onPress={() => this.props.navigation.goBack()}
              color={'#fff'}
              title="Cancel"
            />
          </TouchableHighlight>
        </View>
      )
    }
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
})

export default connect(mapStateToProps)(ModalScreen)


const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#4AA9C5',
    alignItems: 'center',
    justifyContent: 'center',
    padding: 15
  },
  title: {
    fontSize: 42,
    fontWeight: 'bold',
    color: '#fff',
  },
  note: {
    marginBottom: 25,
    fontSize: 14,
    color: '#fff',
    textAlign: 'center'
  },
  suggest: {
    fontWeight: 'bold',
    fontSize: 22,
    marginTop: 20,
    color: '#fff',
    alignSelf: 'flex-start'
  },
  date: {
    width: '95%', 
    backgroundColor: '#fff', 
    borderRadius: 5, 
    marginTop: 5, 
    marginBottom: 5, 
    borderBottomWidth: 0
  },
  input: {
    width: '95%',
    borderRadius: 5, 
    backgroundColor: '#fff', 
    marginTop: 5,
    marginBottom: 5,
  },
  inputContainer: {
    borderBottomWidth: 0
  }, 
  btnContainer: {
    backgroundColor: '#93548F',
    paddingRight: 10,
    paddingLeft: 10,
    borderRadius: 20,
    shadowOffset: {  width: 0,  height: 2 },
    shadowRadius: 5,
    shadowColor: 'black',
    shadowOpacity: .5,
    marginTop: 10,
    width: 200
  }
})

