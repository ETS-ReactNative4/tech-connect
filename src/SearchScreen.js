import React from 'react';
import { StyleSheet, Text, TextInput, View, Button, TouchableHighlight, Image, ScrollView, ActivityIndicator } from 'react-native';
import { connect } from 'react-redux';
import { ButtonGroup } from 'react-native-elements'
import { LinearGradient } from 'expo'
import Connection from './Connection'
import { getAllUsers, getUserInfo, getUsersFilter } from '../apiCalls'
import Icon from 'react-native-vector-icons/Feather'


export class SearchScreen extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      allUsers: [],
      search: '',
      selectedIndex: 0,
      loading: true
    }
  }

  async componentDidMount() {
    const allUsers = await getAllUsers(this.props.user.api_key)
    this.setState({ allUsers, loading: false })
  }

  viewProfile = async (id) => {
    const user = await getUserInfo(id, this.props.user.api_key)
    this.props.navigation.navigate('ProfilePage', {user})
  }

  updateIndex = (selectedIndex) => {
    this.setState({selectedIndex})
  }

  getFilteredUsers = async () => {
    this.setState({ allUsers: [], loading: true })
    const { selectedIndex, search } = this.state
    const queryParams = ['name', 'city', 'position', 'employer']
    const allUsers = await getUsersFilter(this.props.user.api_key, queryParams[selectedIndex], search)
    this.setState({ allUsers, loading: false })
  }

  render() {
    const { allUsers, selectedIndex, search, loading } = this.state
    const userArray = allUsers.map(user => <Connection connection={ user } viewProfile={this.viewProfile.bind(null,user.id)} key={ user.id }/>)
    const buttons = ['Name', 'Location', 'Position', 'Employer']

    return (
      <View style={styles.container}>
        <View style={styles.componentContainer}>
          <Text style={styles.search}>Search by:</Text>
          <ButtonGroup
            onPress={this.updateIndex}
            selectedIndex={selectedIndex}
            selectedButtonStyle={{backgroundColor: '#4AA9C5'}}
            buttons={buttons}
            containerStyle={{height: 30}}
          />
          <View style={styles.searchContainer}>
            <LinearGradient
              colors={['#4AA9C5', '#6364c1', '#93548F']}
              start={0.4}
              style={styles.gradient} >
              <Icon
                name='search'
                size={16}
                color='#FFF'
                onPress={ this.getFilteredUsers }
              />
            </LinearGradient>
            <TextInput
              style={styles.inputContainer}
              placeholder="Search here..."
              underlineColorAndroid="transparent"
              onChangeText={(text) => this.setState({search: text})} 
              value={search}
            />
          </View>
          <ScrollView contentContainerStyle={{alignItems: 'center'}} style={ styles.scrollContainer }>
            <View style={ styles.connectionsContainer }>
              { !loading ? userArray : <ActivityIndicator size="large" color="#fff" /> }
              { 
                (!userArray.length && !loading) && 
                  <Text style={{color: '#fff'}}>No users were found.</Text> 
              }
            </View>
          </ScrollView>
        </View>
      </View>
    )
  }
}

export const mapStateToProps = (state) => ({
  user: state.user
})

export default connect(mapStateToProps)(SearchScreen)


const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'stretch',
    justifyContent: 'center',
    height: '100%',
    width: '100%',
    paddingTop: 20,
    backgroundColor: '#4AA9C5'
  },
  search: {
    marginTop: 10, 
    marginLeft: 10,
    fontSize: 20
  },
  gradient: {
    marginLeft: 4,
    borderRadius: 300, 
    width: 37, 
    height: 37, 
    justifyContent: 'center', 
    alignItems: 'center'
  },
  componentContainer: {
    flex: 1,
    backgroundColor: '#fff', 
    borderRadius: 5,
    justifyContent: 'center', 
    padding: 15,
    alignItems: 'stretch',
    marginTop: 25,
    marginRight: 20,
    marginLeft: 20,
  },
  searchContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 10, 
    borderRadius: 30, 
    borderWidth: 2, 
    borderColor: '#4AA9C5',
    backgroundColor: '#ffffff00', 
    borderBottomWidth: 2, 
  },
  inputContainer: {
    flex: 1,
    fontSize: 16,
    paddingTop: 12,
    paddingBottom: 12,
    paddingLeft: 10,
    backgroundColor: 'transparent'
  },
  scrollContainer: {
    flex: 1,
  },
  innerContainer: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center',
    height: '100%',
    width: '100%',
    backgroundColor: '#FFF',
    paddingBottom: 20,
  },
  connectionsContainer: {
    backgroundColor: '#4AA9C5',
    padding: 10,
    borderRadius: 5,
    width: '100%'
  }
})
