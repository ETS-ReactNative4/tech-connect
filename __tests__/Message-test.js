import React from 'react'
import { shallow } from 'enzyme'
import { Message, mapStateToProps } from '../src/Message'
import { TouchableOpacity, Text } from 'react-native'

describe('Message', () => {
  let wrapper
  let mockMessage

  beforeEach(() => {
    mockMessage = {
      receiver: 'Howard',
      meeting_location: 'Turing',
      sender: 'Howard',
      status: 'Confirmed',
      meeting_date: '2019-02-15T23:11:40.016Z'
    }
    wrapper = shallow(<Message message={ mockMessage } user={{name: 'Joe'}}/>)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render a message based on if you are the sender or receiver', () => {
    wrapper.instance().setState({ showMessage: true})
    expect(wrapper.find(Text).at(3).props().children.join('')).toEqual('Howard has Confirmed a meeting with you at Turing on Fri, 15 Feb 2019 at 23:11:40')
  })

  it('should render a declined message when you are the receiver', () => {
    mockMessage = {
      receiver: 'Joe',
      meeting_location: 'N/A',
      sender: 'Howard',
      status: 'Declined',
      meeting_date: '2019-02-15T23:11:40.016Z'
    }
    wrapper = shallow(<Message message={ mockMessage } user={{name: 'Joe'}}/>)
    wrapper.instance().setState({ showMessage: true})
    expect(wrapper.find(Text).at(3).props().children.join('')).toEqual('You have Declined a meeting with Howard ')
  })

  it('should render a decline message when you are the sender', () => {
    mockMessage = {
      receiver: 'Howard',
      meeting_location: 'N/A',
      sender: 'Joe',
      status: 'Declined',
      meeting_date: '2019-02-15T23:11:40.016Z'
    }
    wrapper = shallow(<Message message={ mockMessage } user={{name: 'Joe'}}/>)
    wrapper.instance().setState({ showMessage: true})
    expect(wrapper.find(Text).at(3).props().children.join('')).toEqual('Howard has Declined a meeting with you ')
  })

  it('should render a confirmed message when you are the receiver', () => {
    mockMessage = {
      receiver: 'Joe',
      meeting_location: 'Turing',
      sender: 'Howard',
      status: 'Confirmed',
      meeting_date: '2019-02-15T23:11:40.016Z'
    }
    wrapper = shallow(<Message message={ mockMessage } user={{name: 'Joe'}}/>)
    wrapper.instance().setState({ showMessage: true})
    expect(wrapper.find(Text).at(3).props().children.join('')).toEqual('You have Confirmed a meeting with Howard at Turing on Fri, 15 Feb 2019 at 23:11:40')
  })

  it('should set state on press of the message and conditionally render a Text element', async () => {
    
    await wrapper.find(TouchableOpacity).simulate('press')

    expect(wrapper.state().showMessage).toEqual(true)
    expect(wrapper.find(Text).length).toEqual(4)

    await wrapper.find(TouchableOpacity).simulate('press')

    expect(wrapper.state().showMessage).toEqual(false)
    expect(wrapper.find(Text).length).toEqual(3)
  })

  describe('mapStateToProps', () => {
    it('should return a user object', () => {
      const mockState = {
        user: { name: 'Kylie' },
        connections: { name: 'Bob' },
        isLoading: false,
        error: ''
      }
      const expected = { user: mockState.user }
      const result = mapStateToProps(mockState)
      expect(result).toEqual(expected)
    })
  })
})