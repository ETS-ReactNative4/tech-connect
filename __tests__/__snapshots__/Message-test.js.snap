// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should match the snapshot 1`] = `
<Component
  style={
    Object {
      "borderBottomColor": "gray",
      "borderBottomWidth": 0.5,
      "display": "flex",
      "flexDirection": "row",
      "padding": 20,
    }
  }
>
  <Component
    style={
      Object {
        "position": "absolute",
        "right": 20,
        "top": 20,
      }
    }
  >
    Invalid Date
  </Component>
  <TouchableOpacity
    activeOpacity={0.2}
    onPress={[Function]}
    style={
      Object {
        "display": "flex",
        "flexDirection": "row",
        "marginRight": 10,
        "maxWidth": "100%",
      }
    }
  >
    <Component
      style={
        Object {
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 2,
          },
          "shadowOpacity": 0.2,
          "shadowRadius": 4,
        }
      }
    >
      <Component
        source={1}
        style={
          Object {
            "borderRadius": 25,
            "height": 50,
            "width": 50,
          }
        }
      />
    </Component>
    <Component
      style={
        Object {
          "marginLeft": 10,
          "marginRight": 30,
          "marginTop": 5,
        }
      }
    >
      <Component
        style={
          Object {
            "fontSize": 18,
          }
        }
      >
        Howard
      </Component>
      <Component
        style={
          Object {
            "color": "gray",
            "marginTop": 5,
          }
        }
      >
        Confirmed
      </Component>
    </Component>
  </TouchableOpacity>
</Component>
`;
